clang/lib/AST/APValue.cpp

```diff
@@ -70,11 +70,13 @@ QualType APValue::LValueBase::getType() const {
     // constexpr int *p = &arr[1]; // valid?
     //
     // For now, we take the most complete type we can find.
-    for (auto *Redecl = cast<ValueDecl>(D->getMostRecentDecl()); Redecl;
+    for (auto *Redecl = cast<ValueDecl>(D->getMostRecentDecl());
+         Redecl && !Redecl->isInvalidDecl();
```

Comment:
This doesn't seem right to me... this change makes us skip the 'right' answer when there is a redecl in the way.  

In reality, I wonder if this whole loop is misguided.  If we want to just find the 'last' non-incomplete-array-type, that search is pretty easy by looping through 'redecls'.  I THINK this ends up better as:

```
QualType T = D->getMostRecentDecl()->getType();
for (const auto *R : D->redecls()) { // unsure if this needs to be 'getMostRecentDecl?'
  if (!R->isInvalidDecl() && !R->getType()->isIncompleteArrayType())
    T = R->getType();
}
return T;
```
`redecl_iterator` is a forward-iterator, else I'd suggest just doing an rbegin/rend type thing on them.

Reply:
We are dealing with ill-formed code here. So there is not "right answer", right? Are you saying that we might misdiagnose an error?

Or are you saying the original code is incorrect and you think there is well-formed code we will not do the correct thing on?

Reply:
The problem is that if the 'invalid' decl is the last one, we end up giving up on this loop entirely.  If the purpose here is to just skip an invalid decl, we probably need to skip JUST it, not all previous ones too.